File: src/main/webapp/ui/src/eln/gallery/galleryVersionHistory.ts
import axios from "@/common/axios";
import { groupByVersion } from "../../util/versionHistory";

/**
 * One audit revision of a Gallery item, as /gallery/ajax/versionHistory returns
 * it. Shaped like the inventory revisions response so that the two share the
 * version-grouping helper.
 */
type ApiRevision = {
  revisionId: number;
  record: {
    version?: number | null;
    lastModified?: string | null;
    modifiedByFullName?: string | null;
    size?: number | null;
    name?: string | null;
    description?: string | null;
  };
};

/** The endpoint's AjaxReturnObject envelope. */
type VersionHistoryResponse = {
  data: { revisions: Array<ApiRevision> } | null;
  error?: { errorMessages?: Array<string> } | null;
};

/**
 * One user-facing version of a Gallery item. Every field describes the version,
 * not the item: see CONTEXT.md, "Version". A null `description` means the
 * audited revision carried none.
 */
export type VersionRow = {
  version: number;
  lastModified: string | null;
  modifiedByFullName: string | null;
  size: number | null;
  name: string | null;
  description: string | null;
};

/**
 * A Gallery item's versions, newest first. Read by the version-history dialog
 * and by the pinned version view. `loadFailedMessage` is passed in so the
 * wording stays with the caller that has a `t`.
 */
export async function fetchVersionHistory(fileId: string, loadFailedMessage: string): Promise<Array<VersionRow>> {
  const { data } = await axios.get<VersionHistoryResponse>(`/gallery/ajax/versionHistory/${fileId}`);
  if (!data.data) {
    // a 200 carrying an error envelope must not read as "no version history yet"
    throw new Error(data.error?.errorMessages?.[0] ?? loadFailedMessage);
  }
  return groupByVersion(data.data.revisions).map(({ version, revision }) => ({
    version,
    lastModified: revision.record.lastModified ?? null,
    modifiedByFullName: revision.record.modifiedByFullName ?? null,
    size: revision.record.size ?? null,
    name: revision.record.name ?? null,
    description: revision.record.description ?? null,
  }));
}

File: src/main/java/com/researchspace/model/dtos/GalleryVersionHistory.java
package com.researchspace.model.dtos;

import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.List;

/**
 * The version history of a Gallery item: every audit revision of it, newest revision id last.
 *
 * <p>Deliberately shaped like the inventory revisions response ({@code
 * ApiInventoryRecordRevisionList}) so that the frontend's shared version-grouping helper consumes
 * both unchanged. See {@code DevDocs/adr/0003-gallery-version-history-endpoint.md}.
 *
 * <p>Callers must group by {@link Item#version()} rather than treat one revision as one version:
 * several revisions can share a version, because not every recorded change bumps the user-facing
 * counter.
 */
public record GalleryVersionHistory(List<Revision> revisions, int revisionsCount) {

  /** One audit revision of a Gallery item. */
  public record Revision(
      long revisionId,
      String revisionType,
      /* named "record" on the wire to match the inventory response; "record" is a restricted
       * identifier in Java, so the component itself is named differently */
      @JsonProperty("record") Item item) {}

  /**
   * The state of the Gallery item at one revision.
   *
   * <p>{@code name} and {@code description} are per-revision, not per-item: uploading a new version
   * can replace the file with a differently named one, and either can be edited at any time. Only
   * the item's identity is shared across versions.
   */
  public record Item(
      Long version,
      String lastModified,
      String modifiedByFullName,
      long size,
      String name,
      String description) {}
}
